src/timers.rs

```diff
@@ -0,0 +1,221 @@
+///! The sampling timer is used for managing ADC sampling and external reference timestamping.
+use super::hal;
+
+macro_rules! timer_channels {
+    ($name:ident, $TY:ident, u32) => {
+        paste::paste! {
+
+            /// The timer used for managing ADC sampling.
+            pub struct $name {
+                timer: hal::timer::Timer<hal::stm32::[< $TY >]>,
+                channels: Option<[< $TY:lower >]::Channels>,
+            }
+
+            impl $name {
+                /// Construct the sampling timer.
+                pub fn new(mut timer: hal::timer::Timer<hal::stm32::[< $TY>]>) -> Self {
+                    timer.pause();
+
+                    Self {
+                        timer,
+                        // Note(unsafe): Once these channels are taken, we guarantee that we do not modify any
+                        // of the underlying timer channel registers, as ownership of the channels is now
+                        // provided through the associated channel structures. We additionally guarantee this
+                        // can only be called once because there is only one Timer2 and this resource takes
+                        // ownership of it once instantiated.
+                        channels: unsafe { Some([< $TY:lower >]::Channels::new()) },
+                    }
+                }
+
+                /// Get the timer capture/compare channels.
+                pub fn channels(&mut self) -> [< $TY:lower >]::Channels {
+                    self.channels.take().unwrap()
+                }
+
+                /// Get the prescaler of a timer.
+                #[allow(dead_code)]
+                pub fn get_prescaler(&self) -> u16 {
+                    let regs = unsafe { &*hal::stm32::$TY::ptr() };
+                    regs.psc.read().psc().bits() + 1
```

Comment:
What if `PSC = 0xffff` (which I think is valid)?

Reply:
These functions are no longer necessary, so I'm going to remove them. This was a bit of a hack early on to gain raw access to the prescaler